vowpalwabbit/global_data.h

```diff
@@ -1,531 +1,535 @@
+  std::array<std::vector<std::shared_ptr<feature_dict>>, NUM_NAMESPACES>
+      namespace_dictionaries{};  // each namespace has a list of dictionaries attached to it
+
+  void (*delete_prediction)(void*);
+  vw_logger logger;
+  bool audit;     // should I print lots of debugging information?
+  bool training;  // Should I train if lable data is available?
+  bool active;
+  bool invariant_updates;  // Should we use importance aware/safe updates
+  uint64_t random_seed;
+  bool random_weights;
+  bool random_positive_weights;  // for initialize_regressor w/ new_mf
+  bool normal_weights;
+  bool tnormal_weights;
+  bool add_constant;
+  bool nonormalize;
+  bool do_reset_source;
+  bool holdout_set_off;
+  bool early_terminate;
+  uint32_t holdout_period;
+  uint32_t holdout_after;
+  size_t check_holdout_every_n_passes;  // default: 1, but search might want to set it higher if you spend multiple
+                                        // passes learning a single policy
+
+  size_t normalized_idx;  // offset idx where the norm is stored (1 or 2 depending on whether adaptive is true)
+
+  uint32_t lda;
+
+  std::string text_regressor_name;
+  std::string inv_hash_regressor_name;
+
+  size_t length() { return ((size_t)1) << num_bits; };
+
+  std::stack<std::tuple<std::string, reduction_setup_fn>> reduction_stack;
+  std::vector<std::string> enabled_reductions;
+
+  // Prediction output
+  std::vector<std::unique_ptr<VW::io::writer>> final_prediction_sink;  // set to send global predictions to.
+  std::unique_ptr<VW::io::writer> raw_prediction;                  // file descriptors for text output.
+
+  VW_DEPRECATED("print has been deprecated, use print_by_ref")
+  void (*print)(VW::io::writer*, float, float, v_array<char>);
+  void (*print_by_ref)(VW::io::writer*, float, float, const v_array<char>&);
+  VW_DEPRECATED("print_text has been deprecated, use print_text_by_ref")
+  void (*print_text)(VW::io::writer*, std::string, v_array<char>);
+  void (*print_text_by_ref)(VW::io::writer*, const std::string&, const v_array<char>&);
+  loss_function* loss;
+
+  VW_DEPRECATED("This is unused and will be removed")
+  char* program_name;
+
+  bool stdin_off;
+
+  bool no_daemon = false;  // If a model was saved in daemon or active learning mode, force it to accept local input when loaded instead.
+
+  // runtime accounting variables.
+  float initial_t;
+  float eta;  // learning rate control.
+  float eta_decay_rate;
+  time_t init_time;
+
+  std::string final_regressor_name;
+
+  parameters weights;
+
+  size_t max_examples;  // for TLC
+
+  bool hash_inv;
+  bool print_invert;
+
+  // Set by --progress <arg>
+  bool progress_add;   // additive (rather than multiplicative) progress dumps
+  float progress_arg;  // next update progress dump multiplier
+
+  std::map<uint64_t, std::string> index_name_map;
+
+  label_type_t label_type;
+
+  polyprediction _predict_buffer;
```

Comment:
where do you see this _predict_buffer landing after `all` struct stops being the one place to dump everything?

Reply:
Good question.  This should live in the "orchestrator".  Since vw is currently the orchestrator, this seems like an appropriate place for it.    In the future if there is a separate orchestrator we should revisit this.

Should probably be private also since the functions using this are now private.  Will do that.